"Exported APIs cannot use un-exported type", even though the type IS exported

## :bug: Bug Report

### General Information
* **JSII Version:** 1.9.0 (build 5c646d5), typescript 3.9.7
* **Platform:** MacOS 19.6.0 Darwin Kernel Version 19.6.0: Sun Jul  5 00:43:10 PDT 2020; root:xnu-6153.141.1~9/RELEASE_X86_64 x86_64

### What is the problem?

I'm having trouble building a package with multiple exports that depend on one another. One exported class uses another one of the same package. Both classes are exported. JSII though fails with *Exported APIs cannot use un-exported type*.

For code reference I point to my PR of iam-floyd where I'm facing this problem https://github.com/udondan/iam-floyd/pull/22/files

My package index looks like this:

```typescript
export { AccessLevel, All, Operator, OperatorModifier } from './shared';
export * from './generated';
export * from './collection';
```

The `./generated` sub-package (do you actually call it like that?) has an [export Ec2](https://github.com/udondan/iam-floyd/pull/22/files#diff-426fd7f5467ee3a487ab7cfce5865b8fR25).

The `./collection` sub-package [uses this Ec2 Class](https://github.com/udondan/iam-floyd/pull/22/files#diff-32c50bdd5713031f5f46089914499ddaR13).

Now when running jsii it complains about Ec2 not being exported:

```
error TS0: Exported APIs cannot use un-exported type iam-floyd.Ec2
```

I also tried to:

- import and use Ec2 directly [instead of the index](https://github.com/udondan/iam-floyd/pull/22/files#diff-32c50bdd5713031f5f46089914499ddaR1): `import {Ec2} from '../generated/amazonec2'`
- export Ec2 in the Collection sub-package, even though it wouldn't be needed outside.
- Add an explicit export to the index.ts: `export { Ec2 } from './generated';`

All the same.

### Verbose Log
```
[2020-07-26T13:04:44.007] [DEBUG] jsii/compiler - Creating or updating /Users/daniel.schroeder/code/iam-floyd/tsconfig.json
[2020-07-26T13:04:52.710] [TRACE] jsii/assembler - Processing source file: lib/index.ts
[2020-07-26T13:04:52.713] [TRACE] jsii/assembler - Processing class: .Collection
[2020-07-26T13:04:52.714] [TRACE] jsii/assembler - Processing method: iam-floyd.Collection#allowEc2InstanceDeleteByOwner
[2020-07-26T13:04:52.735] [INFO] jsii/assembler - Registering JSII class: iam-floyd.Collection
[2020-07-26T13:04:52.737] [ERROR] jsii/compiler - Type model errors prevented the JSII assembly from being created
error TS0: Exported APIs cannot use un-exported type iam-floyd.Ec2
```

Now here's the interesting thing. If I remove the export of `Collection` from the `index.ts` file, it runs through, but the verbose output looks like this:

```
[2020-07-26T13:04:27.755] [DEBUG] jsii/compiler - Creating or updating /Users/daniel.schroeder/code/iam-floyd/tsconfig.json
[2020-07-26T13:04:35.888] [TRACE] jsii/assembler - Processing source file: lib/index.ts
[2020-07-26T13:04:35.892] [TRACE] jsii/assembler - Emitting assembly: /Users/daniel.schroeder/code/iam-floyd/.jsii
```

Even though the `index.ts` contains more exports, I do not see them mentioned at all. To me this looks like something is wrong with detecting the other exports. Though the resulting package works and is perfectly usable. We're using the published package in the latest version w/o issues.

--- a/packages/jsii/lib/assembler.ts
+++ b/packages/jsii/lib/assembler.ts
@@ -750,6 +750,23 @@ export class Assembler implements Emitter {
       return allTypes;
     }
 
+    if (ts.isExportSpecifier(node)) {
+      // This is what happens when one does `export { Symbol } from "./location";`
+      //                   ExportSpecifier:           ~~~~~~
+
+      const resolvedSymbol = this._typeChecker.getExportSpecifierLocalTargetSymbol(
+        node,
+      );
+      if (!resolvedSymbol) {
+        // A grammar error, compilation will already have failed
+        return [];
+      }
+      return this._visitNode(
+        resolvedSymbol.valueDeclaration ?? resolvedSymbol.declarations[0],
+        context,
+      );
+    }
+
     if ((ts.getCombinedModifierFlags(node) & ts.ModifierFlags.Export) === 0) {
       return [];
     }

--- a/packages/jsii/lib/helpers.ts
+++ b/packages/jsii/lib/helpers.ts
@@ -18,18 +18,28 @@ import { loadProjectInfo, ProjectInfo } from './project-info';
  * Compile a piece of source and return the JSII assembly for it
  *
  * Only usable for trivial cases and tests.
+ *
+ * @param source can either be a single `string` (the content of `index.ts`), or
+ *               a map of fileName to content, which *must* include `index.ts`.
  */
 export async function sourceToAssemblyHelper(
-  source: string,
+  source: string | { 'index.ts': string; [name: string]: string },
   cb?: (obj: PackageInfo) => void,
 ): Promise<spec.Assembly> {
+  if (typeof source === 'string') {
+    source = { 'index.ts': source };
+  }
+
   // Easiest way to get the source into the compiler is to write it to disk somewhere.
   // I guess we could make an in-memory compiler host but that seems like work...
   return inTempDir(async () => {
-    const fileName = 'index.ts';
-    await fs.writeFile(fileName, source, { encoding: 'utf-8' });
+    await Promise.all(
+      Object.entries(source).map(([fileName, content]) =>
+        fs.writeFile(fileName, content, { encoding: 'utf-8' }),
+      ),
+    );
     const compiler = new Compiler({
-      projectInfo: await makeProjectInfo(fileName, cb),
+      projectInfo: await makeProjectInfo('index.ts', cb),
     });
     const emitResult = await compiler.emit();
 

--- a/packages/jsii/test/export-specifier.test.ts
+++ b/packages/jsii/test/export-specifier.test.ts
@@ -1,0 +1,16 @@
+import { sourceToAssemblyHelper } from '../lib';
+
+test('export { Foo } from "./foo"', async () => {
+  const assembly = await sourceToAssemblyHelper({
+    'index.ts': 'export { Foo } from "./foo";',
+    'foo.ts': 'export class Foo { private constructor() {} }',
+  });
+
+  expect(assembly.types?.['testpkg.Foo']).toEqual({
+    assembly: 'testpkg',
+    fqn: 'testpkg.Foo',
+    kind: 'class',
+    locationInModule: { filename: 'foo.ts', line: 1 },
+    name: 'Foo',
+  });
+});

